zebra-chain/src/orchard/tree/legacy.rs

```diff
@@ -0,0 +1,102 @@
+//!
+//!
+
+use incrementalmerkletree::{frontier::Frontier, Position};
+
+use super::{Node, NoteCommitmentTree, Root, MERKLE_DEPTH};
+
+///
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename = "NoteCommitmentTree")]
+pub struct LegacyNoteCommitmentTree {
+    inner: LegacyFrontier<Node, MERKLE_DEPTH>,
+    cached_root: std::sync::RwLock<Option<Root>>,
+}
+
+impl From<NoteCommitmentTree> for LegacyNoteCommitmentTree {
+    fn from(nct: NoteCommitmentTree) -> Self {
+        LegacyNoteCommitmentTree {
+            inner: nct.inner.into(),
+            cached_root: nct.cached_root,
+        }
+    }
+}
+
+impl From<LegacyNoteCommitmentTree> for NoteCommitmentTree {
+    fn from(nct: LegacyNoteCommitmentTree) -> Self {
+        NoteCommitmentTree {
+            inner: nct.inner.into(),
+            cached_root: nct.cached_root,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename = "Frontier")]
+struct LegacyFrontier<H, const DEPTH: u8> {
+    frontier: Option<LegacyNonEmptyFrontier<H>>,
+}
+
+impl From<LegacyFrontier<Node, MERKLE_DEPTH>> for Frontier<Node, MERKLE_DEPTH> {
+    fn from(legacy_frontier: LegacyFrontier<Node, MERKLE_DEPTH>) -> Self {
+        if let Some(legacy_frontier_data) = legacy_frontier.frontier {
+            Frontier::from_parts(
+                Position::from(legacy_frontier_data.position.0 as u64),
+                *legacy_frontier_data.leaf.value(),
+                legacy_frontier_data.ommers,
+            )
+            .expect("We should be able to construct a frontier from parts given legacy frontier is not empty")
+        } else {
+            Frontier::empty()
+        }
+    }
+}
+
+impl From<Frontier<Node, MERKLE_DEPTH>> for LegacyFrontier<Node, MERKLE_DEPTH> {
+    fn from(frontier: Frontier<Node, MERKLE_DEPTH>) -> Self {
+        if let Some(frontier_data) = frontier.value() {
+            // We don't have information to build the `Right` side of the `LegacyLeaf` so we just put what we have into the left.
+            let leaf = LegacyLeaf::Left(*frontier_data.leaf());
```

Comment:
Here, if `frontier.position().is_odd()`, then you need to remove the first element from `ommers()` and use that as the left side, and use the leaf as the right side.

Reply:
Got it, thanks! Trying ...

Reply:
It seems there is something else also needed related to the position that need to change. I am not totally sure how it is yet.

Reply:
D'oh. That's right. I have all this working now. Thanks for the help!